httpcore/_async/http2.py

```diff
@@ -578,12 +582,18 @@ async def __aiter__(self) -> typing.AsyncIterator[bytes]:
             # we want to close the response (and possibly the connection)
             # before raising that exception.
             with AsyncShieldCancellation():
-                await self.aclose()
+                # close the stream with cancel
+                await self.aclose(cancel_stream=isinstance(exc, asyncio.exceptions.CancelledError))
```

Comment:
@tomchristie If we don't want to using `_h2_state._get_stream_by_id()` to check if the stream closed before send rst stream, we need to figure out if the stream closed by 'cancel' action.

thus we have to judge the exc type here, and due to the multi backend impls (asyncio, trio, etc.), the condition judgements can be very complex with many `isinstance` calls , it's ugly and hard to maintain afterwards : ( 

so maybe it's more appropriate to just adding a stream closed condition before sending rst stream with CANCEL ? (the only concern is that we called the protected method of h2)

Reply:
ignore this one, found solution by referring the [golang impl](https://github.com/golang/net/blob/4542a42604cd159f1adb93c58368079ae37b3bf6/http2/transport.go#L1639)

judge if the error caused by remote, that's all we need to do : )